src/layers/core/hexagon-layer/hexagon-layer.js

```diff
@@ -0,0 +1,221 @@
+    });
+
+    this.updateRadiusAngle();
+  }
+
+  updateState({props, oldProps, changeFlags: {dataChanged, viewportChanged}} = {}) {
+    const {model, attributeManager} = this.state;
+
+    if (dataChanged) {
+      attributeManager.invalidateAll();
+    }
+
+    // Update the positions in the model if they've changes
+    const verticesChanged =
+      !positionsAreEqual(oldProps.hexagonVertices, props.hexagonVertices);
+
+    if (model && (verticesChanged || viewportChanged)) {
+      this.updateRadiusAngle();
+    }
+
+    this.updateUniforms();
+  }
+
+  updateRadiusAngle() {
+    const {hexagonVertices: vertices, dotRadius} = this.props;
+
+    const vertex0 = vertices[0];
+    const vertex3 = vertices[3];
+
+    // transform to space coordinates
+    const spaceCoord0 = this.projectFlat(vertex0);
+    const spaceCoord3 = this.projectFlat(vertex3);
+
+    // distance between two close centroids
+    const dx = spaceCoord0[0] - spaceCoord3[0];
+    const dy = spaceCoord0[1] - spaceCoord3[1];
+    const dxy = Math.sqrt(dx * dx + dy * dy);
+
+    this.setUniforms({
+      // Calculate angle that the perpendicular hexagon vertex axis is tilted
+      angle: Math.acos(dx / dxy) * -Math.sign(dy) + Math.PI / 2,
+      // Allow user to fine tune radius
+      radius: dxy / 2 * Math.min(1, dotRadius)
+    });
+  }
+
+  getCylinderGeometry(radius) {
+    return new CylinderGeometry({
+      radius,
+      topRadius: radius,
+      bottomRadius: radius,
+      topCap: true,
+      bottomCap: true,
+      height: 1,
+      nradial: 6,
+      nvertical: 1
+    });
+  }
+
+  updateUniforms() {
+    const {opacity, enable3d, elevation, invisibleColor} = this.props;
+    this.setUniforms({
+      opacity,
+      enable3d: enable3d ? 1 : 0,
+      elevation,
+      invisibleColor: [invisibleColor.r, invisibleColor.g, invisibleColor.b]
+    });
+  }
+
+  getShaders() {
+    const vertex = readFileSync(join(__dirname, './hexagon-layer-vertex.glsl'), 'utf8');
+    const lighting = readFileSync(join(__dirname, './lighting.glsl'), 'utf8');
+    const picking = readFileSync(join(__dirname, './picking.glsl'), 'utf8');
+    const vs = picking.concat(lighting).concat(vertex);
+
+    return {
+      vs,
+      fs: readFileSync(join(__dirname, './hexagon-layer-fragment.glsl'), 'utf8')
+    };
+  }
+
+  getModel(gl) {
+    const shaders = assembleShaders(gl, this.getShaders());
+
+    return new Model({
+      gl,
+      id: this.props.id,
+      vs: shaders.vs,
+      fs: shaders.fs,
+      geometry: this.getCylinderGeometry(1),
+      isInstanced: true
+    });
+  }
+
+  draw({uniforms}) {
+    super.draw({uniforms: Object.assign({}, {viewMatrix}, uniforms)});
+  }
+
+  calculateInstancePositions(attribute) {
+    const {data, getCentroid, getElevation} = this.props;
+    const {value, size} = attribute;
+    let i = 0;
+    for (const object of data) {
+      const [lon, lat] = getCentroid(object);
+      const elevation = getElevation(object);
```

Comment:
Note for API audit:
- Another reason to split `positions` into two attributes? I.e. to maintain a one-to-one attribute / accessor mapping.
- Also wonder if `getCentroid` should be `getPosition` (if vertices prop is supplied it may not be the centroid) and if `getElevation` should be `getHeight`